clang/lib/Sema/SemaExprCXX.cpp

```diff
@@ -6027,6 +6027,33 @@ static bool EvaluateBinaryTypeTrait(Sema &Self, TypeTrait BTT, const TypeSourceI
     return cast<CXXRecordDecl>(rhsRecord->getDecl())
       ->isDerivedFrom(cast<CXXRecordDecl>(lhsRecord->getDecl()));
   }
+  case BTT_IsVirtualBaseOf: {
+    const RecordType *BaseRecord = LhsT->getAs<RecordType>();
+    const RecordType *DerivedRecord = RhsT->getAs<RecordType>();
+
+    if (!BaseRecord || !DerivedRecord) {
+      DiagnoseVLAInCXXTypeTrait(Self, Lhs, tok::kw___is_virtual_base_of);
+      DiagnoseVLAInCXXTypeTrait(Self, Rhs, tok::kw___is_virtual_base_of);
+      return false;
+    }
+
+    if (BaseRecord->isUnionType() || DerivedRecord->isUnionType())
+      return false;
+
+    if (!BaseRecord->isStructureOrClassType() ||
+        !DerivedRecord->isStructureOrClassType())
+      return false;
+
+    if (Self.RequireCompleteType(Rhs->getTypeLoc().getBeginLoc(), RhsT,
```

Comment:
What does it mean for there to be a complete derived class type but an incomplete base class type? Shouldn't this require a complete type in both cases to give reasonable diagnostic behavior?

Reply:
The proposal only mentions that the second type has to be complete. My guess is that the reasoning behind this is that in e.g. `__is_virtual_base_of(A, B)`, if `B` is complete but `A` isn’t, then `A` can’t be a base of `B`—since a complete class can’t have an incomplete base—and therefore, it can’t be a virtual base either.

Reply:
I hadn't realized that `is_base_of` already has the same requirement, so I don't insist on a change here. But that said, this presumes users don't have typos they'd like to catch. e.g., 
```
struct Bee { };
struct Be;

struct Derived : virtual Bee {};

static_assert(!__is_virtual_base_of(Be, Derived)); // Oops, I can't spell very well!
```
I think it would be *much* kinder to tell the user that `Be` is incomplete than let them be surprised.

Reply:
> What does it mean for there to be a complete derived class type but an incomplete base class type? Shouldn't this require a complete type in both cases to give reasonable diagnostic behavior?

There is no incomplete base type. We are asking whether some type is a base, and if it isn't, we don't care that it is complete. Type traits don't mandate completeness if they don't need to.